### PR TITLE
Fix TreeNodeObject listener cache

### PR DESCRIPTION
**Pull request by corranwebster**
Fix for #556 and some other fixes to copy-pasted code.

### CODE REVIEW

File: traitsui/tree_node.py [hunk 1]
```diff
@@ -1889,11 +1892,11 @@ def tno_when_label_changed(self, node, listener, remove):
         """
         label = node.label
         if label[:1] != '=':
-            memo = ('label', label, object, listener)
```
Comment (jwiggins) [inline traitsui/tree_node.py]: So it was grabbing the `object` builtin?

Comment (corranwebster) [reply traitsui/tree_node.py]: Yep.  This is a dumb error on my part:

- I assumed `TreeNodeObject` subclassed `TreeNode`
- I then copy-pasted and didn't pay attention to the context fully.

Comment (jwiggins) [reply traitsui/tree_node.py]: I guess the icing on the cake was Mayavi's use of `TreeNodeObject` as the base of it's `Base` class?

Comment (corranwebster) [reply traitsui/tree_node.py]: Yep, whereas I almost never use `TreeNodeObject`.

I've added a regression test.

File: traitsui/tests/editors/test_tree_editor.py [hunk 2]
```diff
@@ -110,6 +153,34 @@ def _test_tree_editor_releases_listeners(self, hide_root, nodes=None,
             notifiers_list = bogus.trait(trait)._notifiers(False)
             self.assertEqual(0, len(notifiers_list))
 
+    def _test_tree_node_object_releases_listeners(self, hide_root, nodes=None,
+                                                  trait='bogus_list',
+                                                  expected_listeners=1):
```
Comment (jwiggins) [inline traitsui/tests/editors/test_tree_editor.py]: None of the calling code uses the `expected_listeners` arg. You should probably remove it.

File: traitsui/tests/editors/test_tree_editor.py [hunk 3]
```diff
@@ -110,6 +153,34 @@ def _test_tree_editor_releases_listeners(self, hide_root, nodes=None,
...
+                bogus=bogus,
+                hide_root=hide_root,
+                nodes=nodes,
+            )
+            ui = tree_editor_view.edit_traits()
+
+            # The TreeEditor sets a listener on the bogus object's
+            # children list
+            notifiers_list = bogus.trait(trait)._notifiers(False)
+            self.assertEqual(expected_listeners, len(notifiers_list))
```
Comment (jwiggins) [inline traitsui/tests/editors/test_tree_editor.py]: And just write `self.assertEqual(1, len(notifiers_list)) here

**Pull request merged**